src/trajectory.jl

```diff
@@ -373,98 +394,102 @@ end
 ## Variants of no-U-turn criteria
 ##
 
-abstract type AbstractTerminationCriterion end
-
 """
 $(TYPEDEF)
-
 Classic No-U-Turn criterion as described in Eq. (9) in [1].
 
 Informally, this will terminate the trajectory expansion if continuing
 the simulation either forwards or backwards in time will decrease the
 distance between the left-most and right-most positions.
 
+# Fields
+$(TYPEDFIELDS)
+
 # References
 1. Hoffman, M. D., & Gelman, A. (2014). The No-U-Turn Sampler: adaptively setting path lengths in Hamiltonian Monte Carlo. Journal of Machine Learning Research, 15(1), 1593-1623. ([arXiv](http://arxiv.org/abs/1111.4246))
 """
-struct ClassicNoUTurn <: AbstractTerminationCriterion end
-
-ClassicNoUTurn(::PhasePoint) = ClassicNoUTurn()
+Base.@kwdef struct ClassicNoUTurn{F<:AbstractFloat} <: DynamicTerminationCriterion
+    max_depth::Int=10
+    Δ_max::F=1000.0
+end
 
 """
 $(TYPEDEF)
-
 Generalised No-U-Turn criterion as described in Section A.4.2 in [1].
 
 # Fields
-
 $(TYPEDFIELDS)
 
 # References
 1. Betancourt, M. (2017). A Conceptual Introduction to Hamiltonian Monte Carlo. [arXiv preprint arXiv:1701.02434](https://arxiv.org/abs/1701.02434).
 """
-struct GeneralisedNoUTurn{T<:AbstractVector{<:Real}} <: AbstractTerminationCriterion
-    "Integral or sum of momenta along the integration path."
-    rho::T
+Base.@kwdef struct GeneralisedNoUTurn{F<:AbstractFloat} <: DynamicTerminationCriterion
+    max_depth::Int=10
+    Δ_max::F=1000.0
 end
 
-GeneralisedNoUTurn(z::PhasePoint) = GeneralisedNoUTurn(z.r)
-
 """
 $(TYPEDEF)
-
 Generalised No-U-Turn criterion as described in Section A.4.2 in [1] with 
 added U-turn check as described in [2].
 
 # Fields
-
 $(TYPEDFIELDS)
 
 # References
 1. Betancourt, M. (2017). A Conceptual Introduction to Hamiltonian Monte Carlo. [arXiv preprint arXiv:1701.02434](https://arxiv.org/abs/1701.02434).
 2. [https://github.com/stan-dev/stan/pull/2800](https://github.com/stan-dev/stan/pull/2800)
 """
-struct StrictGeneralisedNoUTurn{T<:AbstractVector{<:Real}} <: AbstractTerminationCriterion
+Base.@kwdef struct StrictGeneralisedNoUTurn{F<:AbstractFloat} <: DynamicTerminationCriterion
+    max_depth::Int=10
+    Δ_max::F=1000.0
+end
+GeneralisedNoUTurn(tc::StrictGeneralisedNoUTurn) = GeneralisedNoUTurn(tc.max_depth, tc.Δ_max)
+
+struct TurnStatistic{T}
```

Comment:
Maybe considering
```suggestion
struct TurnStatistic{T, C<:AbstractTerminationCriterion}
```
which also removes the need to pass explicit termination criterion types in functions `isterminated(::ClassicNoUTurn, h::Hamiltonian, t::BinaryTree)` below.